zepid/causal/gformula/TimeFixed.py

```diff
@@ -221,3 +221,90 @@ def fit(self, treatment):
         else:  # weighted marginal estimate
             self.marginal_outcome = np.average(g[self.outcome], weights=self.gf[self._weights])
         self.predicted_df = g
+
+    def fit_stochastic(self, p, conditional=None, samples=100, seed=None):
+        """Fits the g-formula for a stochastic intervention. As currently implemented, 'p' percent of the population is
+        randomly treated. This process is repeated 'n' times and the mean is the marginal stochastic outcome
+
+        Parameters
+        ----------
+        p: float, list
+            Percent of the population to randomly treat
+        conditional:
+            List of selection #TODO write this better
+        samples: int, optional
+            Number of resamples to calculate the marginal outcome. Default is 100
+        seed: int, optional
+            Seed for the random process selection
+        """
+        if self._outcome_model is None:
+            raise ValueError('Before the g-formula can be calculated, the outcome model must be specified')
+        if self.exposure_type != 'binary':
+            raise ValueError('Only binary exposures are currently available for the stochastic treatment g-formula')
+        if conditional is not None:
+            if len(p) != len(conditional):
+                raise ValueError("'p' and 'conditional' must be the same length")
+        else:
+            if type(p) != float:
+                raise ValueError("Specified percent must be a float when 'conditional' is not specified")
+
+        if seed is None:
+            pass
+        else:
+            np.random.seed(seed)
+
+        if conditional is not None:
+            self._check_conditional(conditional)
+
+        marginals = []
+        for s in range(samples):
+            g = self.gf.reset_index(drop=True).copy()
+
+            # Selecting out observations
+            if conditional is None:  # Unconditional probabilities
+                if self._weights is None:
+                    treated = np.random.choice(g.index, size=int(p*g.shape[0]), replace=False)
+                else:  # Weighting observations for random treatment
+                    treated = np.random.choice(g.index, size=int(p*g.shape[0]), replace=False,
+                                               p=g[self._weights]/np.sum(g[self._weights]))
+            else:  # Conditional probabilities based on eval()
+                # TODO Checking conditionals are mutually exclusive
+                if self._weights is None:
+                    treated = []
+                    for c, prop in zip(conditional, p):
+                        gs = g.loc[eval(c)].copy()
```

Comment:
Some thoughts on this: it's going to fail if your conditional doesn't use the magic `"g"` variable. Maybe you want that? It's not obvious to the user though. An option to consider, if you haven't already, is the native [`query` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.query.html) in Pandas. That has a) documentation that people can rely on b) probably some cleaner error handling/debugging for users. 

There could be a even better solution that doesn't involve passing in strings, but nothing is obvious to me.

Reply:
Yeah so the magic-g is something baked into custom treatments for both ``TimeFixedGFormula`` and ``TimeVaryGFormula`` but I agree that it is not immediately obvious. It can get confusing in ``TimeVaryGFormula`` (see the [Keil et al 2014 replication](https://github.com/pzivich/zEpid-replications/blob/master/Keil_2014/gformula_keil.ipynb) for how the magic-g is consistently used across the process

``query`` looks to be what I am looking for. I was previous unaware of its existence. I think that is useful

> There could be a even better solution that doesn't involve passing in strings, but nothing is obvious to me.

I haven't found one especially since I would like to retain maximal flexibility in the potential interventions that could be used in the g-formula

Reply:
I've opened #63 for this issue. As it currently stands, all the custom treatments/stochastic treatments use this functionality. I plan on looking into better options. My bandage for now is to make the documentation better and try to handle errors better. ``pd.query`` does not return exactly what I need. ``pd.eval`` looks promising though

This would be updated as part of a more major update (maybe 0.5.0 if ``pd.eval`` is what I want)